cumulusci/core/config/project_config.py

```diff
@@ -516,42 +522,49 @@ def get_flow(self, name):
             flow_config.project_config = self
         return flow_config
 
-    def get_namespace(self, ns):
+    def get_namespace(self, ns: str):
         """Look up another project config by its name in the `sources` config.
 
         Also makes sure the project has been fetched, if it's from an external source.
         """
         spec = getattr(self, f"sources__{ns}")
         if spec is None:
             raise NamespaceNotFoundError(f"Namespace not found: {ns}")
-        return self.include_source(spec)
 
-    def include_source(self, spec):
-        """Make sure a project has been fetched from its source.
+        parsed_spec = None
+        for model_class in [GitHubSourceModel, LocalFolderSourceModel]:
+            try:
+                parsed_spec = model_class(**spec)
+            except ValidationError:
+                pass
+            else:
+                break
 
-        `spec` is a dict which contains different keys depending on the type of source:
+        if not parsed_spec:
+            raise ValueError(f"Invalid source spec: {spec}")
 
-        - `github` indicates a GitHubSource
-        - `path` indicates a LocalFolderSource
+        return self.include_source(parsed_spec)
```

Comment:
include_source is called with a dict (stored on the Step model in the db) in MetaDeploy -- https://github.com/SFDO-Tooling/MetaDeploy/blob/226e32eedd0ba316c3f1f4f43ec774f601485858/metadeploy/api/models.py#L672 -- so it'd be helpful to either keep include_source accepting dict input, or put the dict-to-model logic in its own method that MetaDeploy can use.

Also keep in mind that the old implementation was using the equality of frozendicts to avoid fetching the same source more than once. We need to make sure that if two models are constructed from equivalent dicts, the resulting models are also equivalent.

Reply:
@davisagli See what you think of the tweak I made to support that use case.

As far as caching/equality - the new version is using `HashableBaseModel` so that it can be a dict key. Those models compare based on their field values, and their hash is derived from the JSON representation of those same values. I _think_ the code therefore makes the same guarantees (and there is a test, `test_include_source__cached()` attempting to demonstrate that), but if I've missed a hole please tell me.